File: backend/src/nbplatform/schemas/workspace.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


# ── Workspace ────────────────────────────────────────────────────────────────
class WorkspaceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=10_000)


class WorkspaceUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=10_000)
    is_active: bool | None = None


class WorkspaceGitSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    provider: str
    repo_url: str
    repo_owner: str | None
    repo_name: str | None
    default_branch: str | None
    current_branch: str | None
    last_sync_at: datetime | None


class WorkspaceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    slug: str
    description: str | None
    owner_id: uuid.UUID | None
    root_path: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


class WorkspaceDetail(WorkspaceRead):
    git_repository: WorkspaceGitSummary | None = None


# ── Membros (ACL) ───────────────────────────────────────────────────────────
class WorkspaceMemberRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    user_id: uuid.UUID
    role: Literal["VIEWER", "EDITOR", "OWNER"]
    created_at: datetime


class WorkspaceMemberUpsert(BaseModel):
    role: Literal["VIEWER", "EDITOR", "OWNER"]


# ── File Explorer ────────────────────────────────────────────────────────────
class FileNode(BaseModel):
    name: str
    path: str
    type: Literal["file", "dir"]
    size: int | None = None
    modified_at: float | None = None
    children: list[FileNode] | None = None


class FileContentRead(BaseModel):
    path: str
    kind: Literal["notebook", "text", "binary"]
    content: dict[str, Any] | str | None = None
    etag: str | None = None


class DataPreviewRead(BaseModel):
    columns: list[str]
    dtypes: list[str]
    rows: list[list[Any]]
    total: int | None = None
    truncated: bool = False
    offset: int = 0


class FilePathsRead(BaseModel):
    path: str  # relativo à raiz do Workspace
    name: str
    parent_path: str
    workspace_path: str  # "/<nome do workspace>/<rel>"
    repository_path: str  # rel ao repo git (== path por ora)
    read_example: str | None = None  # snippet pandas/open conforme extensão


class WriteFileRequest(BaseModel):
    # Exatamente um dos dois. `.ipynb` aceita `notebook` (dict) ou `text` (json).
    text: str | None = None
    notebook: dict[str, Any] | None = None


class RenameRequest(BaseModel):
    src: str = Field(min_length=1, alias="from")
    dst: str = Field(min_length=1, alias="to")

    model_config = ConfigDict(populate_by_name=True)


class CopyRequest(RenameRequest):
    pass


class BatchDeleteRequest(BaseModel):
    paths: list[str] = Field(min_length=1, max_length=500)
    recursive: bool = False


class BatchMoveItem(BaseModel):
    src: str = Field(min_length=1, alias="from")
    dst: str = Field(min_length=1, alias="to")

    model_config = ConfigDict(populate_by_name=True)


class BatchMoveRequest(BaseModel):
    items: list[BatchMoveItem] = Field(min_length=1, max_length=500)


class BatchItemResult(BaseModel):
    path: str
    ok: bool
    error: str | None = None
    new_path: str | None = None


class BatchResult(BaseModel):
    results: list[BatchItemResult]


class GenerateFileRequest(BaseModel):
    """Geração de arquivo sintético grande no backend (streaming em disco)."""

    path: str = Field(min_length=1, max_length=1024)
    kind: Literal["csv"] = "csv"
    rows: int = Field(ge=1, le=5_000_000)
    seed: int | None = None


# ── Execução de notebook do Workspace (Fase 2) ───────────────────────────────
class WorkspaceExecuteRequest(BaseModel):
    notebook_path: str = Field(min_length=1, max_length=1024)
    parameters: dict[str, Any] = Field(default_factory=dict)
    idempotency_key: str | None = Field(default=None, max_length=200)

File: frontend/src/lib/workspace.ts
import { apiDelete, apiGet, apiPost, apiPut, getAuthToken } from "@/lib/api";

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? "/api";

// ─── Tipos ──────────────────────────────────────────────────────────────────
export interface Workspace {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  owner_id: string | null;
  root_path: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface WorkspaceGitSummary {
  provider: string;
  repo_url: string;
  repo_owner: string | null;
  repo_name: string | null;
  default_branch: string | null;
  current_branch: string | null;
  last_sync_at: string | null;
}

export interface WorkspaceDetail extends Workspace {
  git_repository: WorkspaceGitSummary | null;
}

export interface FileNode {
  name: string;
  path: string;
  type: "file" | "dir";
  size?: number | null;
  modified_at?: number | null;
  children?: FileNode[] | null;
}

export type FileKind = "notebook" | "text" | "binary";

export interface FileContent {
  path: string;
  kind: FileKind;
  content: Record<string, unknown> | string | null;
  etag?: string | null;
}

// ─── Workspace único (`/root`) ─────────────────────────────────────────────
export const getWorkspace = (): Promise<WorkspaceDetail> => apiGet(`/workspace`);

// ─── File Explorer ─────────────────────────────────────────────────────────
export const getTree = (path = "", depth?: number): Promise<FileNode> => {
  const qs = new URLSearchParams({ path });
  if (depth != null) qs.set("depth", String(depth));
  return apiGet(`/workspace/tree?${qs.toString()}`);
};

export const readFile = (path: string): Promise<FileContent> =>
  apiGet(`/workspace/file?path=${encodeURIComponent(path)}`);

export interface FilePaths {
  path: string;
  name: string;
  parent_path: string;
  workspace_path: string;
  repository_path: string;
  read_example: string | null;
}

export const getFilePaths = (path: string, fromPath?: string): Promise<FilePaths> => {
  const qs = new URLSearchParams({ path });
  if (fromPath) qs.set("from_path", fromPath);
  return apiGet(`/workspace/file/paths?${qs.toString()}`);
};

export const writeFile = (
  path: string,
  body: { text?: string; notebook?: Record<string, unknown>; ifMatch?: string | null },
): Promise<FileContent> =>
  apiPut(
    `/workspace/file?path=${encodeURIComponent(path)}`,
    { text: body.text, notebook: body.notebook },
    body.ifMatch ? { "If-Match": body.ifMatch } : undefined,
  );

export const makeDir = (path: string): Promise<FileNode> =>
  apiPost(`/workspace/dir?path=${encodeURIComponent(path)}`);

export const deleteEntry = (path: string, recursive = false): Promise<void> =>
  apiDelete(`/workspace/file?path=${encodeURIComponent(path)}&recursive=${recursive}`);

export const renameEntry = (from: string, to: string): Promise<FileNode> =>
  apiPost(`/workspace/rename`, { from, to });

export const copyEntry = (from: string, to: string): Promise<FileNode> =>
  apiPost(`/workspace/copy`, { from, to });

export const downloadUrl = (path: string): string =>
  `${API_BASE}/workspace/download?path=${encodeURIComponent(path)}`;

/**
 * Download autenticado (Bearer) via fetch + blob. `window.open(downloadUrl)` não
 * envia o header Authorization e retorna 401 no modo bearer-only.
 */
export async function downloadFile(path: string): Promise<void> {
  const token = getAuthToken();
  const res = await fetch(downloadUrl(path), {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
  });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    let msg = `download → ${res.status}`;
    try {
      msg = (JSON.parse(text) as { error?: { message?: string } })?.error?.message ?? msg;
    } catch {
      /* corpo não-JSON */
    }
    throw new Error(msg);
  }
  const blob = await res.blob();
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  const cd = res.headers.get("Content-Disposition") ?? "";
  const m = /filename="?([^"]+)"?/.exec(cd);
  a.download = m?.[1] ?? path.split("/").pop() ?? "download";
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}

export interface GenerateFileBody {
  path: string;
  rows: number;
  seed?: number | null;
  kind?: "csv";
}

/** Gera um arquivo sintético grande no backend (streaming em disco). */
export const generateFile = (body: GenerateFileBody): Promise<FileNode> =>
  apiPost(`/workspace/generate`, {
    path: body.path,
    rows: body.rows,
    seed: body.seed ?? null,
    kind: body.kind ?? "csv",
  });

export interface WorkspaceExecution {
  id: string;
  status: string;
}

/** Dispara execução de produção (Papermill) de um `.ipynb` do workspace. */
export const executeWorkspaceNotebook = (
  notebookPath: string,
  parameters: Record<string, unknown> = {},
): Promise<WorkspaceExecution> =>
  apiPost(`/workspace/execute`, { notebook_path: notebookPath, parameters });

export async function uploadFile(dir: string, file: File): Promise<FileNode> {
  const form = new FormData();
  form.append("file", file);
  const res = await fetch(`${API_BASE}/workspace/upload?path=${encodeURIComponent(dir)}`, {
    method: "POST",
    headers: getAuthToken() ? { Authorization: `Bearer ${getAuthToken()}` } : {},
    body: form,
  });
  const text = await res.text();
  const parsed = text ? JSON.parse(text) : null;
  if (!res.ok) {
    throw new Error(parsed?.error?.message ?? `upload → ${res.status}`);
  }
  return parsed as FileNode;
}
